src/db/highlights.rs

```diff
@@ -0,0 +1,60 @@
+use super::*;
+use itertools::Itertools;
+
+impl Db {
+    pub async fn get_highlights(&self) -> Result<HashMap<String, Vec<UserId>>> {
+        let mut cache = self.highlight_cache.lock().await;
+        // we have a cache
+        if let Some(cache) = cache.as_ref() {
+            return Ok(cache.clone());
+        // we don't have a cache at all
+        } else {
+            let mut conn = self.pool.acquire().await?;
+            let q = sqlx::query!("select * from highlights")
+                .fetch_all(&mut conn)
+                .await?
+                .into_iter()
+                .map(|x| (x.word, UserId::from(x.user as u64)))
+                .into_group_map();
+            cache.replace(q.clone());
+            return Ok(q.clone());
+        }
+    }
+
+    pub async fn remove_highlight(&self, user: UserId, word: String) -> Result<()> {
+        let mut cache = self.highlight_cache.lock().await;
+        let mut conn = self.pool.acquire().await?;
+        let user = user.0 as i64;
+        sqlx::query!("delete from highlights where word=? and user=?", word, user)
+            .execute(&mut conn)
+            .await?;
+        if let Some(ref mut cache) = cache.as_mut() {
+            if cache.remove(&word).is_none() {
+                bail!("Can't remove a highlight that doesn't exist.")
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn set_highlight(&self, user: UserId, word: String) -> Result<()> {
+        let mut cache = self.highlight_cache.lock().await;
+        let mut conn = self.pool.acquire().await?;
+        let user = user.0 as i64;
+        sqlx::query!(
+            "insert into highlights (word, user) values (?, ?)",
+            word,
+            user
+        )
+        .execute(&mut conn)
+        .await?;
+        let user = user as u64;
+        if let Some(ref mut cache) = cache.as_mut() {
+            if cache.get(&word).is_some() {
+                cache.entry(word).and_modify(|f| f.push(UserId::from(user)));
+            } else {
+                cache.insert(word, vec![UserId::from(user)]);
+            }
```

Comment:
check the entry api https://doc.rust-lang.org/beta/std/collections/hash_map/enum.Entry.html
you can do `cache.entry(word).and_modify(...).or_insert_with(...)` to replace this whole if block